File: src/server_llm/routes_models.py
"""Model selection endpoints."""

from fastapi import APIRouter
from pydantic import BaseModel
from .. import config

router = APIRouter(prefix="/api")

AVAILABLE_MODELS = [
    "gpt-4.1-mini",
    "gpt-4.1",
    "gpt-4o",
    "gpt-4o-mini",
    "o3-mini",
]


# --------------------------------------------------------------
class ModelSwitch(BaseModel):
    """Request to switch model."""
    model: str


# --------------------------------------------------------------
@router.get("/models")
async def get_models():
    """Return current and available models."""
    return {
        "current": config.MODEL_NAME,
        "available": AVAILABLE_MODELS,
    }


# --------------------------------------------------------------
@router.put("/models/current")
async def set_model(req: ModelSwitch):
    """Switch the active model."""
    if req.model not in AVAILABLE_MODELS:
        return {
            "error": (
                f"Unknown model: {req.model}"
            ),
            "available": AVAILABLE_MODELS,
        }
    config.MODEL_NAME = req.model
    return {
        "status": "ok",
        "model": config.MODEL_NAME,
    }

File: server_llm/src/routes_models.rs
/// Model selection endpoints — GET /api/models, PUT /api/models/current
use axum::{routing::{get, put}, Router, Json};
use serde::Deserialize;
use common::config;

const AVAILABLE_MODELS: &[&str] = &[
    "claude-sonnet-4-6",
    "claude-opus-4-6",
    "claude-haiku-4-5-20251001",
];

#[derive(Deserialize)]
pub struct ModelSwitch {
    pub model: String,
}

pub fn router() -> Router {
    Router::new()
        .route("/api/models", get(get_models))
        .route("/api/models/current", put(set_model))
}

async fn get_models() -> Json<serde_json::Value> {
    let current = config::get().model_name.read().unwrap().clone();
    Json(serde_json::json!({
        "current": current,
        "available": AVAILABLE_MODELS,
    }))
}

async fn set_model(Json(req): Json<ModelSwitch>) -> Json<serde_json::Value> {
    if !AVAILABLE_MODELS.contains(&req.model.as_str()) {
        return Json(serde_json::json!({
            "error": format!("Unknown model: {}", req.model),
            "available": AVAILABLE_MODELS,
        }));
    }
    *config::get().model_name.write().unwrap() = req.model.clone();
    Json(serde_json::json!({"status": "ok", "model": req.model}))
}
